### PR TITLE
moved receive from Echo to RawReceiver

### PR DESCRIPTION
**Pull request by PawelA**

### CODE REVIEW

File: libecho/src/IReceiver.h [hunk 1]
```diff
@@ -0,0 +1,5 @@
+class IReceiver {
```
Comment (Riyria-was-taken) [inline libecho/src/IReceiver.h:7]: Please, add header guards

File: libecho/src/RawReceiver.h [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+class RawReceiver : public IReceiver {
+public:
+	RawReceiver(int loFreq, int hiFreq, double magLimitBegin, double magLimitEnd, int winSize);
+	int receive(char *buffer, int size);
+private:
+	void getBytes(char *buffer, int bytes);
+	void clearInput();
+
+	QAudioFormat format;
+	std::unique_ptr<AudioInput> input;
```
Comment (Riyria-was-taken) [inline libecho/src/RawReceiver.h]: You can move it to IReceiver as they are necessary for every Receiver.

File: libecho/src/RawReceiver.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
+#include <QtMultimedia/QAudioFormat>
+#include <memory>
+#include "AudioInput.h"
+#include "IReceiver.h"
+
+class RawReceiver : public IReceiver {
+public:
+	RawReceiver(int loFreq, int hiFreq, double magLimitBegin, double magLimitEnd, int winSize);
```
Comment (Riyria-was-taken) [inline libecho/src/RawReceiver.h]: Maybe add default values for some arguments ^^

File: libecho/src/IReceiver.h (hunk 1 above)
Comment (Riyria-was-taken) [inline libecho/src/IReceiver.h:7]: Don't you need virtual destructor? Also, every receiver needs, for example, windowSize, so maybe add a generic constructor to simplify construction for derived class?

File: libecho/src/IReceiver.h [hunk 4]
```diff
@@ -0,0 +1,5 @@
...
+public:
+	virtual int receive(char *buffer, int size) = 0;
```
Comment (Riyria-was-taken) [inline libecho/src/IReceiver.h]: Don't we use uint8_t as data type we want to operate with?

File: libecho/src/IReceiver.h [hunk 5]
```diff
@@ -0,0 +1,5 @@
...
+private:
+};
```
Comment (Riyria-was-taken) [inline libecho/src/IReceiver.h:23]: We may also need getters and setters for, for example, windowSize, it would be easier to declare them here than in every Receiver class.

File: libecho/src/RawReceiver.h [hunk 6]
```diff
@@ -0,0 +1,22 @@
+#include <QtMultimedia/QAudioFormat>
```
Comment (Riyria-was-taken) [inline libecho/src/RawReceiver.h]: Please, add header guard.

File: libecho/src/RawReceiver.h [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+#include <memory>
+#include "AudioInput.h"
+#include "IReceiver.h"
+
+class RawReceiver : public IReceiver {
+public:
+	RawReceiver(int loFreq, int hiFreq, double magLimitBegin, double magLimitEnd, int winSize);
+	int receive(char *buffer, int size);
```
Comment (Riyria-was-taken) [inline libecho/src/RawReceiver.h]: override?

File: libecho/tests/tmp_test.cpp [hunk 8]
```diff
@@ -1,4 +1,4 @@
-#include "Echo.h"
```
Comment (Riyria-was-taken) [inline libecho/tests/tmp_test.cpp:1]: Instead of commenting out most of the tests you could just leave that include and change only the test you've changed.

File: libecho/src/RawReceiver.cc [hunk 9]
```diff
@@ -0,0 +1,108 @@
+#include <cmath>
+#include "RawReceiver.h"
+
+#define SAMPLE_RATE 44100
```
Comment (Riyria-was-taken) [inline libecho/src/RawReceiver.cc]: static constexpr?

File: libecho/src/RawReceiver.cc [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+    return std::sqrt(re * re + im * im) / samples;
+}
+
+RawReceiver::RawReceiver(int loFreq, int hiFreq, double magLimitBegin, double magLimitEnd, int winSize) {
+    format.setByteOrder(QAudioFormat::LittleEndian);
+    format.setChannelCount(1);
+    format.setCodec("audio/pcm");
+    format.setSampleRate(SAMPLE_RATE);
+    format.setSampleSize(16);
+    format.setSampleType(QAudioFormat::SignedInt);
```
Comment (Riyria-was-taken) [inline libecho/src/RawReceiver.cc]: Maybe move that to the helper method and define arguments as const expressions?

Comment (Riyria-was-taken) [reply libecho/src/RawReceiver.cc]: I mean, something like:
format = getInputFormat();

File: libecho/src/RawReceiver.cc [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+        double val = buffer[i];
+        re += val * std::cos(angle);
+        im += val * std::sin(angle);
+        angle += d_angle;
+    }
+
+    return std::sqrt(re * re + im * im) / samples;
+}
+
+RawReceiver::RawReceiver(int loFreq, int hiFreq, double magLimitBegin, double magLimitEnd, int winSize) {
```
Comment (Riyria-was-taken) [inline libecho/src/RawReceiver.cc]: Initializer list ;)

**Review by Riyria-was-taken — CHANGES_REQUESTED** [review]
Please, run clang-format and clang-tidy, add comments and resolve feedback ^~^
Also, change RawReceiver name to other if we finally decide on one.

File: libecho/src/RawReceiver.h (hunk 3 above)
Comment (PawelA) [reply libecho/src/RawReceiver.h]: I disagree. It doesn't make using this module easier and may cause little, but annoying bugs. Also, these parameters don't have an obvious default.

File: libecho/tests/tmp_test.cpp (hunk 8 above)
Comment (Lorak-mmk) [reply libecho/tests/tmp_test.cpp:1]: Those tests will be deleted anyway - you can just do it now.

File: libecho/src/RawReceiver.cc (hunk 10 above)
Comment (PawelA) [reply libecho/src/RawReceiver.cc]: i don't see how that's a helper method, it doesn't help with anything

File: libecho/src/RawReceiver.cc (hunk 11 above)
Comment (PawelA) [reply libecho/src/RawReceiver.cc]: what?

File: libecho/CMakeLists.txt [hunk 12]
```diff
@@ -25,10 +25,10 @@ set(SRC_FILES
         ${SRC_DIR}/QTInitializer.cc
         ${SRC_DIR}/AudioStream.h
         ${SRC_DIR}/IAudioConverter.h
-        ${SRC_DIR}/BitAudioConverter.h)
-
-set(API_FILES
```
Comment (Lorak-mmk) [inline libecho/CMakeLists.txt:32]: Why did you delete API_FILES variable?

Comment (Lorak-mmk) [reply libecho/CMakeLists.txt:32]: And deleted Echo.h from it?

File: libecho/tests/tmp_test.cpp [hunk 13]
```diff
@@ -35,6 +34,16 @@ TEST(test_send, tmp_test) {
...
 }*/
 
+TEST(test_send_receive, tmp_test) {
+    uint8_t buff[10];
+    BitReceiver receiver(14000, 15000, 600, 6);
+    receiver.receiveFirst(buff, 10);
+    for (int i = 0; i < 10; i++) {
+        printf("%02hhx ", buff[i]);
+    }
+}
```
Comment (Lorak-mmk) [inline libecho/tests/tmp_test.cpp]: This test doesn't really test anything. It also won't be possible to run in CI pipeline. Also, tests should be reproducible.
I'd suggest testing private methods from BitReceiver (by feeding them artificial inputs), preferably by extracting them to some helper class where they are public and can be easily testes, or if that's not possible, you can create test fixture (You can read about text fixture in https://github.com/google/googletest/blob/master/googletest/docs/primer.md ) and set it as friend in BitReceiver to access private methods (I wouldn't recomend going this way).

File: libecho/src/BitReceiver.cc [hunk 14]
```diff
@@ -0,0 +1,206 @@
...
+    res += decodeBit(sync_in.data());
+    res += decodeBit(sync_in.data() + win_size) << 1;
+
+    return res;
+}
+
+int BitReceiver::receiveFirst(uint8_t *buffer, int size) {
+    uint8_t byte = 0;
+
+    input->startStream();
```
Comment (Lorak-mmk) [inline libecho/src/BitReceiver.cc]: Maybe move those starts and stops to constructor/destructor as we discussed, so stream is always running?

**Review by Lorak-mmk — CHANGES_REQUESTED** [review]

File: libecho/tests/tmp_test.cpp (hunk 13 above)
Comment (PawelA) [reply libecho/tests/tmp_test.cpp]: This was just a dummy test to run the syncing and see if it works. I can delete it no problem.
I will test the internal stuff, but not in this PR.

File: libecho/src/RawReceiver.cc (hunk 11 above)
Comment (Lorak-mmk) [reply libecho/src/RawReceiver.cc]: Use initializer list in constructor - instead of `this->loFreq = loFreq;` in constructor body, do something like this:
`RawReceiver::RawReceiver(int loFreq, int hiFreq, double magLimitBegin, double magLimitEnd, int winSize) : loFreq(loFreq) {`

File: libecho/tests/tmp_test.cpp (hunk 13 above)
Comment (Lorak-mmk) [reply libecho/tests/tmp_test.cpp]: Please,delete it

**Review by Lorak-mmk — APPROVED** [review]

File: libecho/src/IReceiver.h [hunk 15]
```diff
@@ -0,0 +1,25 @@
...
+        input->startStream();
+    }
+    virtual ~IReceiver() {
+        input->stopStream();
+    }
+    virtual int receiveFirst(uint8_t *buffer, int size) = 0;
+    virtual int receive(uint8_t *buffer, int size) = 0;
+
+protected:
+    QAudioFormat format;
```
Comment (Riyria-was-taken) [inline libecho/src/IReceiver.h:20]:
```suggestion
    const QAudioFormat format;
```

**Review by Riyria-was-taken — APPROVED** [review]
Thanks ^^
